### PR TITLE
Apply shift in investment variables

### PR DESCRIPTION
**Pull request by fbenke-pik**
This PR is a follow-up to https://github.com/remindmodel/remind/pull/1238

It applies a time shift to the following investment-related variables: `vm_deltaCap`, `vm_costInvTeDir` and `vm_costInvTeAdj`, affecting the variables `New Cap|...`,  `Cumulative Cap|...`, `Energy Investments|...` and `Total Energy costs|...`

Validation:
- Differences in convGDX2MIF can be found here: https://cloud.pik-potsdam.de/index.php/f/22517932
- ~~Differences in convGDX2MIF_LCOE can be found here: https://cloud.pik-potsdam.de/index.php/f/22560628~~
- `piamInterfaces::fixOnRef` did not yield any new isues

### CODE REVIEW

File: R/reportLCOE.R [hunk 1]
```diff
@@ -20,6 +20,8 @@
 #' Can be either "average" (returns only average LCOE),
 #' "marginal" (returns only marginal LCOE), "both" (returns marginal and average LCOE) and
 #' and "marginal detail" (returns table to trace back how marginal LCOE are calculated).
+#' @param gdx_ref a GDX object as created by readGDX, or the path to a gdx of the reference run.
+#' It is used to guarantee consistency for Moving Avg prices before cm_startyear
```
Comment (orichters) [inline R/reportLCOE.R:24]: Really prices here, or is that copied from reportPrices?

File: R/reportCapacity.R [hunk 2]
```diff
@@ -43,13 +48,24 @@ reportCapacity <- function(gdx, regionSubsetList = NULL, t = c(seq(2005, 2060, 5
...
   # data preparation
   ttot <- as.numeric(as.vector(ttot))
-  vm_cap      <- vm_cap[teall2rlf]
-  vm_cap      <- vm_cap[, ttot, ]
+  vm_cap <- vm_cap[teall2rlf]
+  vm_cap <- vm_cap[, ttot, ]
+
   vm_deltaCap <- vm_deltaCap[teall2rlf]
   vm_deltaCap <- vm_deltaCap[, ttot, ]
+
```
Comment (robertpietzcker) [inline R/reportCapacity.R:58]: I would add a comment here to explain the next lines, something like: 

## apply "modifyInvestmentVariables to shift from the model-internal time coverage (deltacap and investment variables for step t represent the average of the years from t-4years to t) to the general convention for the reporting template (all variables represent the average of the years from   t-2.5years to t+2.5years)

File: R/reportEnergyInvestment.R [hunk 3]
```diff
@@ -52,21 +56,37 @@ reportEnergyInvestment <- function(gdx, regionSubsetList = NULL,
...
   pm_data   <- readGDX(gdx, c("pm_data"), format = "first_found")
 
   # data preparation
+
   ttot <- as.numeric(as.vector(readGDX(gdx, "ttot", format = "first_found")))
-  v_directteinv <- v_directteinv[, ttot, ]
-  v_adjustteinv <- v_adjustteinv[, ttot, ]
-  costRatioTdelt2Tdels <- pm_data[, , "inco0.tdelt"] / pm_data[, , "inco0.tdels"]
 
+  if (!is.null(gdx_ref)) {
```
Comment (robertpietzcker) [inline R/reportEnergyInvestment.R:76]: I would add a comment here to explain the next lines, something like: 

apply "modifyInvestmentVariables" to shift from the model-internal time coverage (deltacap and investment variables for step t represent the average of the years from t-4years to t) to the general convention for the reporting template (all variables represent the average of the years from   t-2.5years to t+2.5years)

File: R/reportLCOE.R [hunk 4]
```diff
@@ -138,9 +141,33 @@ reportLCOE <- function(gdx, output.type = "both") {
...
+
+    ## Total direct Investment Cost in Timestep
+    vm_costInvTeDir <- readGDX(gdx, name = c("vm_costInvTeDir", "v_costInvTeDir", "v_directteinv"), field = "l", format = "first_found")[, ttot, ]
+
+    ## total adjustment cost in period
+    vm_costInvTeAdj <- readGDX(gdx, name = c("vm_costInvTeAdj", "v_costInvTeAdj"), field = "l", format = "first_found")[, ttot, ]
+
+    # capacity additions per year
+    vm_deltaCap <- readGDX(gdx, name = c("vm_deltaCap"), field = "l", format = "first_found")[, ttot, ]
+
```
Comment (robertpietzcker) [inline R/reportLCOE.R:154]: I would add a comment here to explain the next lines, something like: 

apply "modifyInvestmentVariables" to shift from the model-internal time coverage (deltacap and investment variables for step t represent the average of the years from t-4years to t) to the general convention for the reporting template (all variables represent the average of the years from   t-2.5years to t+2.5years)

File: R/modifyInvestmentVariables.R [hunk 5]
```diff
@@ -0,0 +1,98 @@
+#' Modify Investment Variables
+#'
+#' Timesteps for investment variables represent a different year range than
```
Comment (robertpietzcker) [inline R/modifyInvestmentVariables.R:3]: I would reformulate slightly:

Years represented by investment variables in the energy system (vm_deltacap, vm_xxx) {please insert all variables this is applied to} are different than in the normal reporting convention. In the current REMIND version, vm_deltacap(t) represents the average of the years t-4years...t (for 5 year time steps), while in the general reporting convention it represents the average of t-2.5years to t+2.5years.

File: R/reportLCOE.R [hunk 6]
```diff
@@ -138,9 +141,33 @@ reportLCOE <- function(gdx, output.type = "both") {
     pm_SEPrice <- readGDX(gdx, "pm_SEPrice", restore_zeros = FALSE)
 
     ## variables
-    vm_costInvTeDir <- readGDX(gdx, name = c("vm_costInvTeDir", "v_costInvTeDir", "v_directteinv"), field = "l", format = "first_found")[, ttot, ] ## Total direct Investment Cost in Timestep
-    vm_costInvTeAdj <- readGDX(gdx, name = c("vm_costInvTeAdj", "v_costInvTeAdj"), field = "l", format = "first_found")[, ttot, ] ## total adjustment cost in period
-    vm_deltaCap   <- readGDX(gdx, name = c("vm_deltaCap"), field = "l", format = "first_found")[, ttot, ]
+
+
+    ## Total direct Investment Cost in Timestep
```
Comment (robertpietzcker) [inline R/reportLCOE.R:146]: I would add a comment here to explain the next lines, something like:

apply "modifyInvestmentVariables" to shift from the model-internal time coverage (deltacap and investment variables for step t represent the average of the years from t-4years to t) to the general convention for the reporting template (all variables represent the average of the years from t-2.5years to t+2.5years)

**Review by robertpietzcker — APPROVED** [review]
thanks a lot for implementing it!

there are many formatting changes that I can't comment on - it seems that the new convention is to add an empty space between function inputs? I like it, but I think if this is now the general formatting it would be good to communicate it as such in the next 1-3 REMIND meetings (to reach all people who are missing in one meeting)

On the concrete shift changes: I made some proposals to make the comments more concrete/ add a comment each time that modifyInvestmentVariables is applied.

I can't comment on the actual calculations, my R is too limited to understand if it was applied correctly - but I trust you :-)

**Comment by fbenke-pik** [comment]
> there are many formatting changes that I can't comment on - it seems that the new convention is to add an empty space between function inputs? I like it, but I think if this is now the general formatting it would be good to communicate it as such in the next 1-3 REMIND meetings (to reach all people who are missing in one meeting)

The formatting changes are not mandatory (they wont stop you from finishing buildLibrary). I just applied the auto formatting and fixed indentation. The former can be done by using `lucode2::autoFormat(PATH_TO_FILE)`, the latter is done in RStudio by entering Ctrl + I. Both are automatic steps that improve readability a bit, so they don't hurt, but are optional. Sorry for mixing them with the other changes, as it makes it harder to check the PR, but I did not make any changes other than formatting on existing code in the report functions, so focusing on the new functions and the new code blocks is sufficient.

**Comment by fbenke-pik** [comment]
> I can't comment on the actual calculations, my R is too limited to understand if it was applied correctly - but I trust you :-)

Take a look at the before/after comparison and check if you spot some irregularities
Differences in convGDX2MIF can be found here: https://cloud.pik-potsdam.de/index.php/f/22517932
Differences in convGDX2MIF_LCOE can be found here: https://cloud.pik-potsdam.de/index.php/f/22560628

**Comment by robertpietzcker** [comment]
> Differences in convGDX2MIF_LCOE can be found here: https://cloud.pik-potsdam.de/index.php/f/22560628

thanks for reminding me about this again. 

The capacity and energy investment variables look fine; the LCOE variables don't. 

the average LCOE is calculated based on the total output of the technology divided by the total costs incurred until that moment. 
with the new shift of investments by half a time step in the reporting, this means the model sees already lots of investments while not actually producing any energy, leading to hugely overblown numbers for technologies that are growing. 

I would propose to revert all the changes in the LCOE script.

or what do you say, @fschreyer ?

**Review by fschreyer — CHANGES_REQUESTED** [review]
Thanks, Falk! That's great. I checked the ``modifyInvestmentVariables`` function with an example GDX and the key part looks all correct to me. I made the comments a bit more verbose and opened a PR to your branch. 

I agree with Robert that this function should not be applied in the reportLCOE function. Here, the current model definition of investments over t-4 to t of timestep t is actually what we want in the calculation of average LCOE.

File: R/modifyInvestmentVariables.R [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  if (!is.null(ref)) {
+    joinedYears <- intersect(getYears(x, as.integer = TRUE), getYears(ref, as.integer = TRUE))
+    fixedYears <- joinedYears[joinedYears < startYear]
+    if (length(fixedYears) == 0) {
+      return(x)
+    }
+    ref <- modifyInvestmentVariables(ref)
+    joinedNames <- intersect(getNames(x), getNames(ref))
+    joinedRegions <- intersect(getItems(ref, dim = 1), getItems(x, dim = 1))
+    x[joinedRegions, fixedYears, joinedNames] <- ref[joinedRegions, fixedYears, joinedNames]
```
[the remark this answers is not in the document]
Comment (fbenke-pik) [reply R/modifyInvestmentVariables.R:96]: Ok, I can implement it that way. It was written like this, as I adapted the logic from Oliver's implementation for Prices.

File: R/reportCapitalStock.R [hunk 8]
```diff
@@ -3,136 +3,155 @@
...
+    tmp2 <- mbind(tmp2, setNames(dimSums(vm_deltaCap[, , "apCarElT"] * pm_conv_cap_2_MioLDV, dim = c(3.1, 3.2)), "Est EV LDV Sales (million vehicles)"))
+    tmp2 <- mbind(tmp2, setNames(dimSums(vm_deltaCap[, , "apCarH2T"] * pm_conv_cap_2_MioLDV, dim = c(3.1, 3.2)), "Est H2 LDV Sales (million vehicles)"))
+    tmp2 <- mbind(tmp2, setNames(dimSums(vm_deltaCap[, , "apCarPeT"] * pm_conv_cap_2_MioLDV, dim = c(3.1, 3.2)), "Est ICE LDV Sales (million vehicles)"))
+
+    tmp2 <- mbind(tmp2, setNames(dimSums(vm_deltaCap[, , c("apCarDiT", "apcarDiEffT", "apcarDiEffH2T")], dim = c(3.1, 3.2)), "Services and Products|Transport|non-LDV|Sales|uedit (arbitrary unit)"))
+    tmp2 <- mbind(tmp2, setNames(dimSums(vm_deltaCap[, , "apCarDiT"], dim = c(3.1, 3.2)), "Services and Products|Transport|non-LDV|Sales|apCarDiT (arbitrary unit)"))
+    tmp2 <- mbind(tmp2, setNames(dimSums(vm_deltaCap[, , "apcarDiEffT"], dim = c(3.1, 3.2)), "Services and Products|Transport|non-LDV|Sales|apcarDiEffT (arbitrary unit)"))
+    tmp2 <- mbind(tmp2, setNames(dimSums(vm_deltaCap[, , "apcarDiEffH2T"], dim = c(3.1, 3.2)), "Services and Products|Transport|non-LDV|Sales|apcarDiEffH2T (arbitrary unit)"))
```
[the remark this answers is not in the document]
Comment (robertpietzcker) [reply R/reportCapitalStock.R:109]: yes, I agree

**Pull request merged**